File: backend/app/schemas/flywheel.py
"""Schemas for the flywheel learning loop — classification & extraction feedback."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class ClassificationFeedbackCreate(BaseModel):
    """Record a classification correction."""

    document_id: str
    predicted_type: str
    corrected_type: str

    model_config = ConfigDict(from_attributes=True)


class ExtractionFeedbackCreate(BaseModel):
    """Record an extraction correction or confirmation."""

    document_id: str
    field_name: str
    predicted_value: str
    corrected_value: str | None = None
    accepted: bool = False

    model_config = ConfigDict(from_attributes=True)


class PerTypeAccuracy(BaseModel):
    accuracy: float
    total: int
    correct: int


class ConfusionPair(BaseModel):
    predicted: str
    actual: str
    count: int


class AccuracyMetrics(BaseModel):
    """Classification or extraction accuracy metrics."""

    overall_accuracy: float
    total_predictions: int | None = None
    total_extractions: int | None = None
    total_corrections: int
    per_type_accuracy: dict[str, PerTypeAccuracy] | None = None
    per_field_accuracy: dict[str, PerTypeAccuracy] | None = None
    confusion_matrix: list[ConfusionPair] | None = None
    trend: str  # improving | stable | declining

    model_config = ConfigDict(from_attributes=True)


class LearnedRule(BaseModel):
    """A rule discovered from correction patterns."""

    predicted_type: str
    corrected_type: str
    occurrence_count: int
    confidence: float
    suggestion: str

    model_config = ConfigDict(from_attributes=True)


class FlywheelDashboardRead(BaseModel):
    """Aggregated flywheel dashboard metrics."""

    classification_accuracy: float
    extraction_accuracy: float
    total_documents_processed: int
    total_corrections: int
    correction_rate: float
    top_confusion_pairs: list[ConfusionPair]
    learned_rules_count: int
    learned_rules: list[LearnedRule]
    improvement_trend: str  # improving | stable | declining

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/flywheel.ts
import { apiClient } from '@/api/client';

export interface ClassificationFeedbackCreate {
  document_id: string;
  predicted_type: string;
  corrected_type: string;
}

export interface ExtractionFeedbackCreate {
  document_id: string;
  field_name: string;
  predicted_value: string;
  corrected_value?: string | null;
  accepted: boolean;
}

export interface ConfusionPair {
  predicted: string;
  actual: string;
  count: number;
}

export interface PerTypeAccuracy {
  accuracy: number;
  total: number;
  correct: number;
}

export interface ClassificationAccuracy {
  overall_accuracy: number;
  per_type_accuracy: Record<string, PerTypeAccuracy>;
  confusion_matrix: ConfusionPair[];
  total_predictions: number;
  total_corrections: number;
  trend: string;
}

export interface ExtractionAccuracy {
  overall_accuracy: number;
  per_field_accuracy: Record<string, PerTypeAccuracy>;
  total_extractions: number;
  total_corrections: number;
  trend: string;
}

export interface LearnedRule {
  predicted_type: string;
  corrected_type: string;
  occurrence_count: number;
  confidence: number;
  suggestion: string;
}

export interface FlywheelDashboard {
  classification_accuracy: number;
  extraction_accuracy: number;
  total_documents_processed: number;
  total_corrections: number;
  correction_rate: number;
  top_confusion_pairs: ConfusionPair[];
  learned_rules_count: number;
  learned_rules: LearnedRule[];
  improvement_trend: string;
}

export const flywheelApi = {
  recordClassificationFeedback: async (data: ClassificationFeedbackCreate) => {
    const response = await apiClient.post('/flywheel/classification-feedback', data);
    return response.data;
  },

  recordExtractionFeedback: async (data: ExtractionFeedbackCreate) => {
    const response = await apiClient.post('/flywheel/extraction-feedback', data);
    return response.data;
  },

  getClassificationAccuracy: async (orgId?: string): Promise<ClassificationAccuracy> => {
    const params = orgId ? { org_id: orgId } : {};
    const response = await apiClient.get<ClassificationAccuracy>('/flywheel/accuracy/classification', { params });
    return response.data;
  },

  getExtractionAccuracy: async (orgId?: string): Promise<ExtractionAccuracy> => {
    const params = orgId ? { org_id: orgId } : {};
    const response = await apiClient.get<ExtractionAccuracy>('/flywheel/accuracy/extraction', { params });
    return response.data;
  },

  getLearnedRules: async (documentType?: string): Promise<LearnedRule[]> => {
    const params = documentType ? { document_type: documentType } : {};
    const response = await apiClient.get<LearnedRule[]>('/flywheel/learned-rules', { params });
    return response.data;
  },

  getDashboard: async (orgId?: string): Promise<FlywheelDashboard> => {
    const params = orgId ? { org_id: orgId } : {};
    const response = await apiClient.get<FlywheelDashboard>('/flywheel/dashboard', { params });
    return response.data;
  },
};
